File: v4/rust/iprange-livedb/src/fixed_tree/delete.rs
//! Fixed-tree deletion without occupancy rebalancing.

use crate::error::{Error, Result};
use crate::mapping::ByteSource;
use crate::page_io::PageEdit;
use crate::slotted_page;

use super::insert::{propagate_first, propagate_first_from};
use super::page::{branch_child, parse};
use super::{first_key, private_path, Codec, Path, RetiredPages, Store};

pub(super) struct Target {
    pub(super) path: Path,
    pub(super) page_number: u32,
    pub(super) header: crate::slotted_page::Header,
    pub(super) index: usize,
}

#[cfg(test)]
pub(crate) fn delete<C: Codec, S: Store>(
    store: &mut S,
    root: &mut u32,
    key: C::Key,
    retired: &mut RetiredPages,
) -> Result<bool> {
    if !super::read::contains::<C, S>(store, *root, key)? {
        return Ok(false);
    }
    delete_existing::<C, S>(store, root, key, retired)?;
    Ok(true)
}

pub(crate) fn delete_existing<C: Codec, S: Store>(
    store: &mut S,
    root: &mut u32,
    key: C::Key,
    retired: &mut RetiredPages,
) -> Result<()> {
    let target = locate::<C, S>(store, root, key, retired)?;
    delete_target::<C, S>(store, root, target)
}

pub(super) fn delete_target<C: Codec, S: Store>(
    store: &mut S,
    root: &mut u32,
    target: Target,
) -> Result<()> {
    if target.header.item_count > 1 {
        remove_leaf_record::<C, S>(store, root, target)?;
        return Ok(());
    }
    store.discard_private(target.page_number)?;
    remove_empty_child::<C, S>(store, root, &target.path)
}

fn locate<C: Codec, S: Store>(
    store: &mut S,
    root: &mut u32,
    key: C::Key,
    retired: &mut RetiredPages,
) -> Result<Target> {
    let leaf = private_path::<C, S>(store, root, key, retired)?;
    let (index, exists) = leaf.selection;
    if !exists {
        return Err(Error::Corrupt("B+tree key disappeared during deletion"));
    }
    Ok(Target {
        path: leaf.path,
        page_number: leaf.page_number,
        header: leaf.header,
        index,
    })
}

fn remove_leaf_record<C: Codec, S: Store>(
    store: &mut S,
    root: &mut u32,
    target: Target,
) -> Result<()> {
    store.update_page(target.page_number, |page| {
        remove_at::<C, _>(page, &target.header, target.index)
    })?;
    if target.index != 0 {
        return Ok(());
    }
    let first = first_key::<C, S>(store, target.page_number, 0)?;
    propagate_first::<C, S>(store, root, &target.path, first)
}

fn remove_empty_child<C: Codec, S: Store>(
    store: &mut S,
    root: &mut u32,
    path: &Path,
) -> Result<()> {
    let mut depth = path.depth;
    if depth == 0 {
        *root = 0;
        return Ok(());
    }

    while depth > 0 {
        depth -= 1;
        let frame = path.frames[depth];
        let target_txn = store.target_txn();
        let header = store.inspect_page(frame.page_number, |page| {
            parse::<C, _>(page, target_txn, None)
        })?;
        if header.item_count == 1 {
            store.discard_private(frame.page_number)?;
            if depth == 0 {
                *root = 0;
                return Ok(());
            }
            continue;
        }

        store.update_page(frame.page_number, |page| {
            remove_at::<C, _>(page, &header, frame.index)
        })?;
        let output_count = header.item_count - 1;
        if depth == 0 && output_count == 1 {
            let child = store.inspect_page(frame.page_number, |page| {
                let output = parse::<C, _>(page, target_txn, Some(header.level))?;
                branch_child::<C, _>(page, &output, 0, store.page_limit())
            })?;
            *root = child;
            store.discard_private(frame.page_number)?;
            return Ok(());
        }
        if frame.index == 0 {
            let first = first_key::<C, S>(store, frame.page_number, header.level)?;
            propagate_first_from::<C, S>(store, root, path, depth, first)?;
        }
        return Ok(());
    }
    Ok(())
}

fn remove_at<C: Codec, D: PageEdit>(
    page: &mut D,
    header: &crate::slotted_page::Header,
    index: usize,
) -> Result<()> {
    let old_len = super::page::codec_cell::<C, _>(page.view(), header, index)?.len();
    slotted_page::remove(page, header, index, old_len)
}

File: v4/go/internal/tree/delete.go
// Fixed-tree deletion without occupancy rebalancing (Rust
// fixed_tree/delete.rs).

package tree

import (
	"github.com/firehol/iprange/v4/go/internal/format"
)

// Following carries the first surviving leaf value after a removed run.
type Following[T any] struct {
	Key  Key
	Leaf T
}

// RemovedRun reports the outcome of one leaf-run removal.
type RemovedRun[T any] struct {
	Removed   uint64
	Following *Following[T]
}

// Target identifies one deletion position in a private leaf.
type Target struct {
	Path       Path
	PageNumber uint32
	Header     Header
	Index      int
}

// DeleteExisting deletes the existing key from the tree (Rust
// delete_existing); the caller must have proven the key exists.
func DeleteExisting[T any](codec Codec[T], store Store, root *uint32, key Key, retired RetiredPages) (RetiredPages, error) {
	leaf, retired, err := PrivatePath(codec, store, root, key, retired)
	if err != nil {
		return RetiredPages{}, err
	}
	if !leaf.Exists {
		return RetiredPages{}, corrupt("B+tree key disappeared during deletion")
	}
	return retired, DeleteTarget(codec, store, root, &Target{Path: leaf.Path, PageNumber: leaf.PageNumber, Header: leaf.Header, Index: leaf.Index})
}

// RemoveLeafRun removes every consecutive leaf record from index of the
// key's private leaf while the include predicate holds, mirroring Rust
// remove_leaf_run. Fixed-size leaf cells are required. include is a
// direct function parameter: the caller's closure stays on the stack.
func RemoveLeafRun[T any](codec Codec[T], store Store, root *uint32, key Key, include func(T) (bool, error)) (RemovedRun[T], error) {
	leaf, retired, err := PrivatePath(codec, store, root, key, RetiredPages{})
	if err != nil {
		return RemovedRun[T]{}, err
	}
	if retired.Len() != 0 {
		return RemovedRun[T]{}, corrupt("private B+tree run retired a page")
	}
	if !leaf.Exists {
		return RemovedRun[T]{}, corrupt("B+tree run start key is missing")
	}
	index := leaf.Index
	var following *Following[T]
	end := index
	for end < int(leaf.Header.ItemCount) {
		cell, err := codecCell(codec, leaf.Page, &leaf.Header, end)
		if err != nil {
			return RemovedRun[T]{}, err
		}
		item, err := codec.ReadLeaf(cell)
		if err != nil {
			return RemovedRun[T]{}, err
		}
		keep, err := include(item)
		if err != nil {
			return RemovedRun[T]{}, err
		}
		if !keep {
			key, err := codec.ReadKey(cell, 0)
			if err != nil {
				return RemovedRun[T]{}, err
			}
			following = &Following[T]{Key: key, Leaf: item}
			return RemovedRun[T]{Removed: 0, Following: following}, nil
		}
		end++
	}
	if end == index {
		return RemovedRun[T]{Removed: 0, Following: following}, nil
	}
	if following == nil {
		if adjacent, found, err := adjacentLeaf(codec, store, leaf.Path, AdjacentAfter); err != nil {
			return RemovedRun[T]{}, err
		} else if found {
			following = &Following[T]{Key: adjacent.key, Leaf: adjacent.leaf}
		}
	}
	removed := uint64(end - index)
	if end-index == int(leaf.Header.ItemCount) {
		if err := store.DiscardPrivate(leaf.PageNumber); err != nil {
			return RemovedRun[T]{}, err
		}
		if err := removeEmptyChild(codec, store, root, &leaf.Path); err != nil {
			return RemovedRun[T]{}, err
		}
	} else {
		cellLen, ok := FixedCellSize(codec, 0)
		if !ok {
			return RemovedRun[T]{}, unsupported("B+tree run removal requires fixed leaf cells")
		}
		page, tag, err := store.Update(leaf.PageNumber)
		if err != nil {
			return RemovedRun[T]{}, err
		}
		if _, err := format.SlottedRemoveFixedRange(page, &leaf.Header, index, end-index, cellLen); err != nil {
			return RemovedRun[T]{}, err
		}
		if err := store.RestoreDirty(leaf.PageNumber, tag); err != nil {
			return RemovedRun[T]{}, err
		}
		if index == 0 {
			first, err := FirstKey(codec, store, leaf.PageNumber, 0)
			if err != nil {
				return RemovedRun[T]{}, err
			}
			if err := PropagateFirst(codec, store, root, &leaf.Path, first); err != nil {
				return RemovedRun[T]{}, err
			}
		}
	}
	return RemovedRun[T]{Removed: removed, Following: following}, nil
}

// DeleteTarget deletes one record from a private leaf, collapsing the tree
// when the leaf or an ancestor becomes empty (Rust delete_target).
func DeleteTarget[T any](codec Codec[T], store Store, root *uint32, target *Target) error {
	if target.Header.ItemCount > 1 {
		return removeLeafRecord(codec, store, root, target)
	}
	if err := store.DiscardPrivate(target.PageNumber); err != nil {
		return err
	}
	return removeEmptyChild(codec, store, root, &target.Path)
}

func removeLeafRecord[T any](codec Codec[T], store Store, root *uint32, target *Target) error {
	page, tag, err := store.Update(target.PageNumber)
	if err != nil {
		return err
	}
	if err := removeAt(codec, page, &target.Header, target.Index); err != nil {
		return err
	}
	if err := store.RestoreDirty(target.PageNumber, tag); err != nil {
		return err
	}
	if target.Index != 0 {
		return nil
	}
	first, err := FirstKey(codec, store, target.PageNumber, 0)
	if err != nil {
		return err
	}
	return PropagateFirst(codec, store, root, &target.Path, first)
}

func removeEmptyChild[T any](codec Codec[T], store Store, root *uint32, path *Path) error {
	depth := path.Depth()
	if depth == 0 {
		*root = 0
		return nil
	}
	for depth > 0 {
		depth--
		frame := path.Frame(depth)
		targetTxn := store.TargetTxn()
		page, err := store.Inspect(frame.PageNumber)
		if err != nil {
			return err
		}
		header, err := parse(codec, page, targetTxn, 0, false)
		if err != nil {
			return err
		}
		if header.ItemCount == 1 {
			if err := store.DiscardPrivate(frame.PageNumber); err != nil {
				return err
			}
			if depth == 0 {
				*root = 0
				return nil
			}
			continue
		}
		page, tag, err := store.Update(frame.PageNumber)
		if err != nil {
			return err
		}
		if err := removeAt(codec, page, &header, frame.Index); err != nil {
			return err
		}
		if err := store.RestoreDirty(frame.PageNumber, tag); err != nil {
			return err
		}
		outputCount := int(header.ItemCount) - 1
		if depth == 0 && outputCount == 1 {
			page, err := store.Inspect(frame.PageNumber)
			if err != nil {
				return err
			}
			output, err := parse(codec, page, targetTxn, header.Level, true)
			if err != nil {
				return err
			}
			child, err := branchChild(codec, page, &output, 0, store.PageLimit())
			if err != nil {
				return err
			}
			*root = child
			return store.DiscardPrivate(frame.PageNumber)
		}
		if frame.Index == 0 {
			first, err := FirstKey(codec, store, frame.PageNumber, header.Level)
			if err != nil {
				return err
			}
			if err := propagateFirstFrom(codec, store, root, path, depth, first); err != nil {
				return err
			}
		}
		return nil
	}
	return nil
}

func removeAt[T any](codec Codec[T], page []byte, header *Header, index int) error {
	oldCell, err := codecCell(codec, page, header, index)
	if err != nil {
		return err
	}
	return format.SlottedRemove(page, header, index, len(oldCell))
}
